### DISCUSSION TITLE
fast api requires higher version of typing extension(>=4.8.0) where as tensorflow requires lower version of typing extension(<4.6.0)

### DISCUSSION THREAD

**Discussion by shubham-gupta0** [fastapi/fastapi#11746] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import numpy as np
import tensorflow as tf
from tensorflow.keras.models import load_model
from tensorflow.keras.preprocessing.image import img_to_array
from PIL import Image
from io import BytesIO
from fastapi import FastAPI, UploadFile, File, HTTPException
from pydantic import BaseModel

app = FastAPI()

# Load the model once at startup
model = None

def preprocess_image(image_file, target_size=(128, 128)):
    try:
        image = Image.open(BytesIO(image_file))
        image = image.convert('RGB')  # Convert to RGB
        image = image.resize(target_size, Image.LANCZOS)
        image = img_to_array(image)
        image = tf.keras.applications.mobilenet_v2.preprocess_input(image)
        return image
    except Exception as e:
        print(f"Error processing image: {e}")
        return None

def load_mobilenet_model(model_path):
    global model
    model = load_model(model_path)

def compare_images(image1_file, image2_file):
    image1 = preprocess_image(image1_file)
    image2 = preprocess_image(image2_file)
    
    if image1 is None or image2 is None:
        print("Error in preprocessing images.")
        return None, None

    image1 = np.expand_dims(image1, axis=0)  # Add batch dimension
    image2 = np.expand_dims(image2, axis=0)  # Add batch dimension

    prediction = model.predict([image1, image2])

    # Assuming the model outputs a distance metric where smaller values indicate greater similarity
    distance = prediction[0][0]

    # Determine if images are of the same muzzle or not based on a threshold
    threshold = 0.5  # You may need to adjust this threshold based on your model's performance
    if distance > threshold:
        result = "Same muzzle"
    else:
        result = "Different muzzles"

    return result, float(distance)  # Convert numpy.float32 to native Python float

@app.on_event("startup")
def load_model_on_startup():
    model_path = "muzzle_model.h5"
    load_mobilenet_model(model_path)

class CompareMuzzlesResponse(BaseModel):
    result: str
    distance: float

@app.post("/compare-muzzles", response_model=CompareMuzzlesResponse)
async def compare_muzzles(file1: UploadFile = File(...), file2: UploadFile = File(...)):
    try:
        image1_file = await file1.read()
        image2_file = await file2.read()
        result, distance = compare_images(image1_file, image2_file)
        if result is None:
            raise HTTPException(status_code=400, detail="Error processing images")
        return CompareMuzzlesResponse(result=result, distance=distance)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app, host="0.0.0.0", port=8000)
```

### Description

i was using the above code to run my machine learning model i tried it hosting to my ubuntu 20.04 server in which i got conflicting errors when installed the required libraries is was mainly due to typing extentions version requirements by the other libraries
the error when i installed typing extension 4.5.0:
ERROR: torch 2.3.1 has requirement typing-extensions>=4.8.0, but you'll have typing-extensions 4.5.0 which is incompatible.
ERROR: pydantic 2.7.4 has requirement typing-extensions>=4.6.1, but you'll have typing-extensions 4.5.0 which is incompatible.
ERROR: pydantic-core 2.18.4 has requirement typing-extensions!=4.7.0,>=4.6.0, but you'll have typing-extensions 4.5.0 which is incompatible.
ERROR: fastapi 0.111.0 has requirement typing-extensions>=4.8.0, but you'll have typing-extensions 4.5.0 which is incompatible.

the errors when i used typing extension of 4.8.0:
ERROR: tensorflow 2.13.1 has requirement typing-extensions<4.6.0,>=3.6.6, but you'll have typing-extensions 4.12.2 which is incompatible.

### Operating System

Other

### Operating System Details

OS : ubuntu 20.04

### FastAPI Version

0.111.0

### Pydantic Version

2,7,4

### Python Version

3.8.10

**Comment by mattmess1221** [comment]
If you use pdm, you can override it with the [`[tool.pdm.resolution.overrides]`](https://pdm-project.org/en/latest/usage/config/#override-the-resolved-package-versions) table in `pyproject.toml`.

```toml
[tool.pdm.resolution.overrides]
typing-extensions = ">=4.8.0"
```

This *should* be safe since typing-extensions doesn't have breaking changes.

**Comment by YuriiMotov** [comment]
Seems that it's outdated. No problem in current versions
